### PR TITLE
Shrink `ast::{Generics,Impl,Block}`

### PR DESCRIPTION
**Pull request by nnethercote**
r? @ghost

### CODE REVIEW

File: compiler/rustc_ast/src/ast.rs [hunk 1]
```diff
@@ -2440,7 +2440,7 @@ pub struct ForeignMod {
 
 #[derive(Clone, Encodable, Decodable, Debug)]
 pub struct EnumDef {
-    pub variants: Vec<Variant>,
+    pub variants: Box<[Variant]>,
```
Comment (Kobzol) [inline compiler/rustc_ast/src/ast.rs:2443]: Maybe we could use `std:.boxed::ThinBox` or something similar to further reduce the size from `16` to `8` bytes? By storing the slice size on the heap.

Comment (nnethercote) [reply compiler/rustc_ast/src/ast.rs:2443]: There is also `thin_vec::ThinVec`, which I introduced to the compiler in #100869. Unfortunately, getting any of these kinds of changes to show a clear performance benefit has been hard.

**Comment by nnethercote** [comment]
This one seems unlikely to produce any wins.

**Pull request closed without merging**